src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1100,37 +1132,227 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
+        /// <exception cref="DataApiBuilderException">Throws exception when an authorization check fails.</exception>
+        private void AuthorizeEntityAndFieldsForMutation(
+            IMiddlewareContext context,
+            string clientRole,
+            string entityName,
+            EntityActionOperation operation,
+            string inputArgumentName,
+            IDictionary<string, object?> parametersDictionary
+            )
+        {
+            IInputField schemaForArgument = context.Selection.Field.Arguments[inputArgumentName];
+
+            // Dictionary to store a mapping from entityName to all field names being referenced from that entity in the mutation.
+            Dictionary<string, HashSet<string>> entityAndFieldsToAuthorize = new();
+            object? parameters;
+            if (parametersDictionary.TryGetValue(inputArgumentName, out parameters))
+            {
+                // Get all the entity names and field names referenced in the mutation.
+                PopulateMutationFieldsToAuthorize(entityAndFieldsToAuthorize, schemaForArgument, entityName, context, parameters, _runtimeConfigProvider.GetConfig());
+            }
+
+            // List of entities which are being referenced in the mutation.
+            IEnumerable<string> entityNames = entityAndFieldsToAuthorize.Keys;
+
+            // Perform authorization check at entity level.
+            foreach(string entityNameInMutation in entityNames)
             {
-                throw new DataApiBuilderException(
-                    message: "Unauthorized due to one or more fields in this mutation.",
-                    statusCode: HttpStatusCode.Forbidden,
-                    subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
-                );
+                if (!_authorizationResolver.AreRoleAndOperationDefinedForEntity(entityNameInMutation, clientRole, operation))
+                {
+                    throw new DataApiBuilderException(
+                        message: $"The client has insufficient permissions to perform the operation on the entity: {entityNameInMutation}",
+                        statusCode: HttpStatusCode.Forbidden,
+                        subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed
+                        );
+                }
+            }
+
+            // Perform authorization checks at field level.
+            foreach ((string entityNameInMutation, HashSet<string> fieldsInEntity) in entityAndFieldsToAuthorize)
+            {
+                if (!AreFieldsAuthorizedForEntity(clientRole, entityNameInMutation, EntityActionOperation.Create, fieldsInEntity))
+                {
+                    throw new DataApiBuilderException(
+                        message: $"Access is forbidden to one or more fields in the entity: {entityNameInMutation} referenced in this mutation.",
```

Comment:
Why cant we maintain the same exception message as we had previously for point Create mutation? And still have it for update, delete => "Unauthorized due to one or more fields in this mutation."?

Reply:
we can. I'll incorporate this.